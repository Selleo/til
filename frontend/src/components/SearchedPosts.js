import React from 'react'
import { useSelector } from 'react-redux'
import SearchedPhrase from './SearchedPhrase'
import PostsList from '../components/PostsList'
import NothingFound from './NothingFound'
import { useSearchedPosts } from '../utils/customHooks/useSearchedPosts'
import PostSkeletonTemplate from './PostSkeletonTemplate'

const SearchedPosts = () => {
  const posts = useSearchedPosts()
  const searchQuery = useSelector(({ searchQuery }) => searchQuery)

  if (!posts) {
    if (searchQuery) {
      return <NothingFound text={searchQuery} />
    } else {
      return <PostSkeletonTemplate />
    }
  }

  return (
    <>
      <SearchedPhrase phrase={searchQuery} />
      <PostsList posts={posts} />
    </>
  )
}

export default SearchedPosts
