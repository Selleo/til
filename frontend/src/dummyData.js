export const initialState = {
  categories: [],
  currentUser: null,
  users: [],
  posts: []
};

export const user = {
  email: "l.walczak@selleo.com",
  firstName: null,
  image:
    "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
  lastName: null,
  uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
};

export const allUsers = [
  {
    email: "l.walczak@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
  },
  {
    email: "maciekPlacek@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAIdsdfjdkjdfi/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-dadlfj33343443"
  }
];

export const categories = [
  { id: 1, name: "java" },
  { id: 2, name: "javascript" }
];

export const posts = [
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adsfdf",
    categoriesIds: [2, 1],
    id: 1,
    isPublic: false,
    likes: [
      {
        post_id: 1,
        user_uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
      }
    ],
    likesCount: 1,
    title: "title "
  },
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adfefsddf",
    categoriesIds: [3],
    id: 2,
    isPublic: false,
    likes: [],
    likesCount: 0,
    title: "maciekdfd"
  }
];
