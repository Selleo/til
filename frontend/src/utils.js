import { useLocation } from "react-router-dom";
const localStorageKey = "til_token";

// FETCH POSTS
export const fetchData = async url => {
  let optionsToken = null;

  if (getToken()) {
    optionsToken = {
      headers: {
        Authorization: `Bearer ${getToken()}`,
        "Content-Type": "application/json"
      }
    };
  }

  const response = await fetch(url, optionsToken);
  const data = await response.json();

  return data;
};

export const fetchSinglePost = async (url, id) => {
  let optionsToken = null;

  if (getToken()) {
    optionsToken = {
      headers: {
        Authorization: `Bearer ${getToken()}`,
        "Content-Type": "application/json"
      }
    };
  }

  const response = await fetch(url + id, optionsToken);
  const post = response.json();

  return post;
};

export const fetchReviewPost = async url => {
  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.json();
};

// APPROVE POST
export const approvePost = async url => {
  const response = await fetch(url, {
    method: "PUT",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.ok;
};

// ADD POST
export const postData = async (url, data) => {
  const response = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    },
    body: data
  });

  return response.ok;
};

// UPDATE POST
export const updateData = async (url, data) => {
  const response = await fetch(url, {
    method: "PATCH",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    },
    body: data
  });

  return response.ok;
};

// DELETE POST
export const deleteData = async (url, data) => {
  const response = await fetch(url, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.ok;
};

// FETCH USERS POSTS
export const fetchUserPosts = async (url, id) => {
  try {
    const response = await fetch(url + id);
    const user = await response.json();
    const userPosts = user.posts;

    return userPosts;
  } catch (error) {
    console.warn(error);
  }
};

// LIKES
export const handleLike = async (postId, method) => {
  const response = await fetch(`api/posts/${postId}/likes`, {
    method: `${method}`,
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.ok;
};

export const checkHasLiked = (likes, userId) => {
  const likesArray = likes.map(like => like.user_uuid);

  return likesArray.includes(userId);
};

// AUTHENTICATION
export const isAuthenticated = () => {
  const token = getToken();
  if (token) return true;

  return false;
};

// TOKENS
export const deleteToken = () => {
  window.localStorage.removeItem(localStorageKey);
};

export const getToken = () => {
  return window.localStorage.getItem(localStorageKey);
};

// URLS
export const getCurrentURL = () => {
  return document.URL;
};

export const useQuery = () => {
  const queryString = useLocation().search;

  return new URLSearchParams(queryString);
};

// FETCH USER
export const fetchUser = async url => {
  const response = await fetch(url, {
    method: "GET",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.json();
};

// HELPERS
export const convertToSelectOptions = data => {
  return data.map(el => ({ value: el.id, label: el.name }));
};

// REACTIONS
export const handleReaction = async (postId, method, reaction) => {
  let ending = "";

  if (method === "DELETE") {
    ending = `/${reaction}`;
  }

  const response = await fetch(
    `api/posts/${postId}/reactions${ending}`,
    {
      method: `${method}`,
      headers: {
        Authorization: `Bearer ${getToken()}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({ type: reaction })
    }
  );

  return response.ok;
};

export const convertReactions = reactions => {
  const reactionsType = ["like", "funny", "love", "surprised"];

  let filteredReactions = [];

  reactionsType.forEach(reaction => {
    const result = reactions.filter(el => el.type === reaction);
    const reactionObj = { type: reaction, whoReacted: result };
    filteredReactions.push(reactionObj);
  });

  return filteredReactions;
};

export const checkHasReacted = (reaction, userId) => {
  const reactionArray = reaction.map(reaction => reaction.user_uuid);

  return reactionArray.includes(userId);
};
