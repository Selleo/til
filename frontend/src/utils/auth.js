const localStorageKey = "til_token";

// TOKENS
export const deleteToken = () => {
  window.localStorage.removeItem(localStorageKey);
};

export const getToken = () => {
  return window.localStorage.getItem(localStorageKey);
};

// AUTHENTICATION
export const isAuthenticated = () => {
  const token = getToken();
  if (token) {
    return true;
  }

  return false;
};
