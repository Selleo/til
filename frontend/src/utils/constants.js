export const statusType = {
  idle: 'idle',
  loading: 'loading',
  fetched: 'fetched',
}
