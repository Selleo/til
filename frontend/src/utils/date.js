const monthNames = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
]

export const getDate = () => {
  const today = new Date()
  let minutes = today.getMinutes()
  if (minutes < 10) {
    minutes = `0${today.getMinutes()}`
  }
  const time = `${today.getHours()}:${minutes}`
  const month = monthNames[today.getMonth() - 1].slice(0, 3)
  const date = `${month} ${today.getDate()} | ${time}`

  return date
}
