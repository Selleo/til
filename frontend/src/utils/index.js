export {
  request,
  fetchData,
  fetchSinglePost,
  fetchUserPosts
} from "./posts/CRUDposts";

export { approvePost } from "./reviewPost/approvePost";
export { fetchReviewPost } from "./reviewPost/fetchReviewPost";

export { getCurrentURL, useQuery } from "./url/getURLparams";

export { handleReaction } from "./posts/handleReactions";

export { fetchUser } from "./dataFetching/fetchUser";

export {
  convertToSelectOptions,
  convertReactions,
  checkHasReacted
} from "./array/helpers";

export { isAuthenticated, deleteToken } from "./auth";
