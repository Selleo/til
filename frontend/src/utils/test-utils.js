import React from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import configureMockStore from "redux-mock-store";
import { initialState as defaultState } from "../dummyData";

export const WrapperWithStoreAndRouter = ({
  children,
  initialState = defaultState,
}) => {
  const history = createMemoryHistory();
  const mockStore = configureMockStore([]);
  const store = mockStore(initialState);

  return (
    <Provider store={store}>
      <Router history={history}>{children}</Router>
    </Provider>
  );
};
export const WrapperWithStore = ({ children, initialState = defaultState }) => {
  const mockStore = configureMockStore([]);
  const store = mockStore(initialState);

  return <Provider store={store}>{children}</Provider>;
};

export const WrapperWithRouter = ({ children }) => {
  const history = createMemoryHistory();

  return <Router history={history}>{children}</Router>;
};
