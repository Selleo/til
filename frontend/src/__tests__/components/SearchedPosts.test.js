import React from "react";
import { render } from "@testing-library/react";
import SearchedPosts from "../../components/SearchedPosts";
import { WrapperWithStoreAndRouter } from "../../utils/test-utils";
import { posts as searchedPosts } from "../../dummyData";

describe("<SearchedPosts/>", () => {
  it("renders correctly with posts", () => {
    const { container } = render(
      <WrapperWithStoreAndRouter initialState={{ searchedPosts }}>
        <SearchedPosts />
      </WrapperWithStoreAndRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("returns `nothing found` when no posts founded", () => {
    const { getByText } = render(
      <WrapperWithStoreAndRouter initialState={{ searchedPosts: [] }}>
        <SearchedPosts />
      </WrapperWithStoreAndRouter>
    );
    const message = getByText(/nothing found/i);

    expect(message).toBeInTheDocument();
  });
});
