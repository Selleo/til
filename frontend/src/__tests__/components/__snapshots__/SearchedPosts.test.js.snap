// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchedPosts/> renders correctly with posts 1`] = `
<div>
  <div
    class="posts"
  >
    <section
      class="post"
    >
      <article>
        <a
          href="/posts/1"
        >
          <h1>
            title 
          </h1>
        </a>
        <p>
          written by 
           
        </p>
        <p>
          adsfdf
        </p>
      </article>
      <button>
        Copy post Url
      </button>
      <p>
        like
        : 
        0
      </p>
      <p>
        funny
        : 
        0
      </p>
      <p>
        love
        : 
        0
      </p>
      <p>
        surprised
        : 
        0
      </p>
    </section>
    <section
      class="post"
    >
      <article>
        <a
          href="/posts/2"
        >
          <h1>
            maciekdfd
          </h1>
        </a>
        <p>
          written by 
           
        </p>
        <p>
          adfefsddf
        </p>
      </article>
      <button>
        Copy post Url
      </button>
      <p>
        like
        : 
        0
      </p>
      <p>
        funny
        : 
        0
      </p>
      <p>
        love
        : 
        0
      </p>
      <p>
        surprised
        : 
        0
      </p>
    </section>
  </div>
</div>
`;
